src/libmeasurement_kit/ooni/templates.cpp

```diff
@@ -93,38 +93,51 @@ void http_request(Var<Entry> entry, Settings settings, http::Headers headers,
     http::request(
         settings, headers, body,
         [=](Error error, Var<http::Response> response) {
-            Entry rr;
-            /*
-             * Note: `probe_ip` comes from an external service, hence
-             * we MUST call `represent_string` _after_ `redact()`.
-             */
-            for (auto pair: headers) {
-                rr["request"]["headers"][pair.first] =
-                    represent_string(redact(pair.second));
-            }
-            rr["request"]["body"] = represent_string(redact(body));
-            rr["request"]["url"] = settings.at("http/url").c_str();
-            rr["request"]["method"] = settings.at("http/method").c_str();
 
-            // XXX we should probably update OONI data format to remove this.
-            rr["method"] = settings.at("http/method").c_str();
+            auto dump = [&](Var<http::Response> response) {
+                Entry rr;
+
+                if (!!error) {
+                    rr["failure"] = error.as_ooni_error();
+                } else {
+                    rr["failure"] = nullptr;
+                }
+
+                if (!!response) {
+                    /*
+                     * Note: `probe_ip` comes from an external service, hence
+                     * we MUST call `represent_string` _after_ `redact()`.
+                     */
+                    for (auto pair : response->headers) {
+                        rr["response"]["headers"][pair.first] =
+                            represent_string(redact(pair.second));
+                    }
+                    rr["response"]["body"] =
+                        represent_string(redact(response->body));
+                    rr["response"]["response_line"] =
+                        represent_string(redact(response->response_line));
+                    rr["response"]["code"] = response->status_code;
+
+                    auto request = response->request;
+                    for (auto pair : request->headers) {
+                        rr["request"]["headers"][pair.first] =
+                            represent_string(redact(pair.second));
+                    }
+                    rr["request"]["body"] =
+                        represent_string(redact(request->body));
+                    rr["request"]["url"] = request->url.str();
+                    rr["request"]["method"] = request->method;
+                }
+                return rr;
+            };
 
-            if (!error) {
-                for (auto pair: response->headers) {
-                    rr["response"]["headers"][pair.first] =
-                        represent_string(redact(pair.second));
+            if (!!response) {
+                for (Var<http::Response> x = response; !!x; x = x->previous) {
+                    (*entry)["requests"].push_back(dump(x));
```

Comment:
I've just run a web connectivity with input `https://httpbin.org/redirect/3` and this is an extract from the result NJSON file:

```JSON
{
    "requests": [
        {
            "failure": null,
            "request": {
                "body": "",
                "headers": {
                    "Accept": "text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8",
                    "Accept-Language": "en-US;q=0.8,en;q=0.5",
                    "User-Agent": "Mozilla/5.0 (Windows NT 6.1) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/47.0.2526.106 Safari/537.36"
                },
                "method": "GET",
                "url": "https://httpbin.org/get"
            },
            "response": {
                "body": "{\n  \"args\": {}, \n  \"headers\": {\n    \"Accept\": \"text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8\", \n    \"Accept-Language\": \"en-US;q=0.8,en;q=0.5\", \n    \"Host\": \"httpbin.org\", \n    \"User-Agent\": \"Mozilla/5.0 (Windows NT 6.1) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/47.0.2526.106 Safari/537.36\"\n  }, \n  \"origin\": \"[REDACTED]\", \n  \"url\": \"https://httpbin.org/get\"\n}\n",
                "code": 200,
                "headers": {
                    "Access-Control-Allow-Credentials": "true",
                    "Access-Control-Allow-Origin": "*",
                    "Connection": "keep-alive",
                    "Content-Length": "385",
                    "Content-Type": "application/json",
                    "Date": "Tue, 28 Feb 2017 16:02:35 GMT",
                    "Server": "nginx"
                },
                "response_line": "HTTP/1.1 200 OK"
            }
        },
        {
            "failure": null,
            "request": {
                "body": "",
                "headers": {
                    "Accept": "text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8",
                    "Accept-Language": "en-US;q=0.8,en;q=0.5",
                    "User-Agent": "Mozilla/5.0 (Windows NT 6.1) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/47.0.2526.106 Safari/537.36"
                },
                "method": "GET",
                "url": "https://httpbin.org/relative-redirect/1"
            },
            "response": {
                "body": "",
                "code": 302,
                "headers": {
                    "Access-Control-Allow-Credentials": "true",
                    "Access-Control-Allow-Origin": "*",
                    "Connection": "keep-alive",
                    "Content-Length": "0",
                    "Content-Type": "text/html; charset=utf-8",
                    "Date": "Tue, 28 Feb 2017 16:02:35 GMT",
                    "Location": "/get",
                    "Server": "nginx"
                },
                "response_line": "HTTP/1.1 302 FOUND"
            }
        },
        {
            "failure": null,
            "request": {
                "body": "",
                "headers": {
                    "Accept": "text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8",
                    "Accept-Language": "en-US;q=0.8,en;q=0.5",
                    "User-Agent": "Mozilla/5.0 (Windows NT 6.1) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/47.0.2526.106 Safari/537.36"
                },
                "method": "GET",
                "url": "https://httpbin.org/relative-redirect/2"
            },
            "response": {
                "body": "",
                "code": 302,
                "headers": {
                    "Access-Control-Allow-Credentials": "true",
                    "Access-Control-Allow-Origin": "*",
                    "Connection": "keep-alive",
                    "Content-Length": "0",
                    "Content-Type": "text/html; charset=utf-8",
                    "Date": "Tue, 28 Feb 2017 16:02:34 GMT",
                    "Location": "/relative-redirect/1",
                    "Server": "nginx"
                },
                "response_line": "HTTP/1.1 302 FOUND"
            }
        },
        {
            "failure": null,
            "request": {
                "body": "",
                "headers": {
                    "Accept": "text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8",
                    "Accept-Language": "en-US;q=0.8,en;q=0.5",
                    "User-Agent": "Mozilla/5.0 (Windows NT 6.1) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/47.0.2526.106 Safari/537.36"
                },
                "method": "GET",
                "url": "https://httpbin.org/redirect/3"
            },
            "response": {
                "body": "<!DOCTYPE HTML PUBLIC \"-//W3C//DTD HTML 3.2 Final//EN\">\n<title>Redirecting...</title>\n<h1>Redirecting...</h1>\n<p>You should be redirected automatically to target URL: <a href=\"/relative-redirect/2\">/relative-redirect/2</a>.  If not click the link.",
                "code": 302,
                "headers": {
                    "Access-Control-Allow-Credentials": "true",
                    "Access-Control-Allow-Origin": "*",
                    "Connection": "keep-alive",
                    "Content-Length": "247",
                    "Content-Type": "text/html; charset=utf-8",
                    "Date": "Tue, 28 Feb 2017 16:02:34 GMT",
                    "Location": "/relative-redirect/2",
                    "Server": "nginx"
                },
                "response_line": "HTTP/1.1 302 FOUND"
            }
        }
    ]
}
```